CLUSTER.md

```diff
@@ -0,0 +1,451 @@
+replication offset matches the one provided by the primary it simply announces itself as the new primary, and the former
+primary will set itself as replica.
+
+If a node does not get a heartbeat response for a long enough time period it will fence itself off and stop serving traffic.
+
+![Failover](images/failover.jpg?raw=true "Failover")
+
+#### Failover Coordinator Tick
+
+In order to avoid relying on multiple clocks when making failover decisions, the node relies on a monotonically increasing
+counter maintained by the Failover Coordinator. Every time a heartbeat is received by the Failover Coordinator, it is saved internally with the current tick
+(which is increased by the Failover Coordinator every K milliseconds). Since a node knows what tick its heartbeat was received in, it can tell how
+many ticks have passed since the primary sent its last heartbeat.
+
+### Resharding
+
+When a Topology Director receives an update which moves a set of slot to a specific shard it sets the appropriate slots to the
+importing/migrating state and bumps the topology epoch. This new slot topology will eventually reach all the relevant nodes.  
+
+It is expected that we will have atomic slot migration (based on fork+serialize like Redis replication) in Redis 
+_before_ Flotilla is GA - in which case Flotilla will only support that and not the existing key migration flow.
+Failing that, the behavior around key migration will be exactly same as in the current Redis Cluster - A node with a slot 
+set to Migrating returns ASK as appropriate, a node with a slot set to Importing requires Asking (and that behavior can
+be enhanced with things like bulk/full slot migrate without any impact on the Flotilla design).
+
+In our reference implementation we currently require the administrator to explicitly call the Topology Director with another update to terminate
+the Migrating/Importing state, but an alternative would be to handle this inside Flotilla by having nodes with migrating
+slots add the number of keys per slot to their heartbeat -- once that number drops to zero the Topology Director can automatically clear
+the state (It's also possible to emit the MIGRATE_DONE command from the primary migrating shard).
+
+![Resharding](images/resharding.jpg?raw=true "Failover")
+
+## Flotilla Implementation Details
+
+<!--
+Maybe mention Flotilla as a reference architecture and different possible implementations and an existing prototype implementation.
+-->
+
+### Roles
+
+While the Flotilla spec mostly discusses data nodes (and not Failover Coordinator & Topology Director implementation), in our reference implementation
+nodes within a Flotilla deployment have one or more roles:
+- Topology Director
+- Failover Coordinator
+- Data node
+
+The role of the node is specified in the join request (see [TBD Topics](#tbd-topics)).
+
+### Node and Failover Coordinator IDs
+
+All IDs in the system are:
+- Unique numbers
+- Monotonically increasing
```

Comment:
>Monotonically increasing

Can you expand a bit on the rationale? I am curious to know what use cases could be enabled by this strong property? This can't be provided today on cluster V1 where we don't run consensus for any ID assignment.

Reply:
done